File: Algorithms/src/main/java/com/cuongd/study/algsp1/book/ch2/QuickSort.java
package com.cuongd.study.algsp1.book.ch2;

import edu.princeton.cs.algs4.StdRandom;

import java.lang.reflect.InvocationTargetException;

public class QuickSort extends SortCommon {
    @Override
    public void sort(Comparable[] a) {
        StdRandom.shuffle(a);
        sort(a, 0, a.length - 1);
    }

    private static void sort(Comparable[] a, int lo, int hi) {
        if (hi <= lo) return;
        int j = partition(a, lo, hi);
        sort(a, lo, j-1);
        sort(a, j+1, hi);
    }

    private static int partition(Comparable[] a, int lo, int hi) {
        Comparable key = a[lo];
        int i = lo, j = hi + 1;

        while (true) {
            while (less(a[++i], key)) if (i == hi) break;
            while (less(key, a[--j])) if (j == lo) break;
            if (i >= j) break;
            exchange(a, i, j);
        }
        exchange(a, lo, j);
        return j;
    }

    public static void main(String[] args) throws InvocationTargetException, NoSuchMethodException,
            InstantiationException, IllegalAccessException {
        run(QuickSort.class);
    }
}

File: Introduction to the Design and Analysis of Algorithms/src/ch5/quicksort.c
#include <assert.h>

void swap(int a[], int i, int j) {
  int t = a[i];

  a[i] = a[j];
  a[j] = t;
}

int partition(int a[], int l, int r) {
  int p = a[l], i = l, j = r;

  while (i < j) {
    while (a[++i] < p)
      ;
    while (a[--j] > p)
      ;
    swap(a, i, j);
  }
  swap(a, i, j);
  swap(a, l, j);
  return j;
}

void quicksort_(int a[], int l, int r) {
  int s;

  if (l < r) {
    s = partition(a, l, r);
    quicksort_(a, l, s);
    quicksort_(a, s + 1, r);
  }
}

void quicksort(int a[], int n) { quicksort_(a, 0, n); }

void assert_xs(int xs[], int ys[], int n) {
  int i;

  quicksort(xs, n);
  for (i = 0; i < n; i++) {
    assert(xs[i] == ys[i]);
  }
}

int main() {
  int a[] = {3, 1, 5, 4, 7, 6, 2};
  int aa[] = {1, 2, 3, 4, 5, 6, 7};

  assert_xs(a, aa, 7);
}
